src/com/amazon/ion/impl/IonCursorBinary.java

```diff
@@ -178,8 +262,362 @@ class IonCursorBinary implements IonCursor {
+        }
+    }
+
+    /**
+     * Validates the given configuration.
+     * @param configuration the configuration to validate.
+     * @return the validated configuration.
+     */
+    private static IonBufferConfiguration validate(IonBufferConfiguration configuration) {
+        if (configuration.getInitialBufferSize() < 1) {
+            throw new IllegalArgumentException("Initial buffer size must be at least 1.");
+        }
+        if (configuration.getMaximumBufferSize() < configuration.getInitialBufferSize()) {
+            throw new IllegalArgumentException("Maximum buffer size cannot be less than the initial buffer size.");
+        }
+        return configuration;
+    }
+
+    /**
+     * Constructs a refillable cursor from the given input stream.
+     * @param configuration the configuration to use.
+     * @param alreadyRead the byte array containing the bytes already read (often the IVM).
+     * @param alreadyReadOff the offset into `alreadyRead` at which the first byte that was already read exists.
+     * @param alreadyReadLen the number of bytes already read from `alreadyRead`.
+     */
+    IonCursorBinary(
+        IonBufferConfiguration configuration,
+        InputStream inputStream,
+        byte[] alreadyRead,
+        int alreadyReadOff,
+        int alreadyReadLen
+    ) {
+        this.dataHandler = (configuration == null) ? null : configuration.getDataHandler();
+        if (configuration == null) {
+            if (inputStream instanceof ByteArrayInputStream) {
+                // ByteArrayInputStreams are fixed-size streams. Clamp the reader's internal buffer size at the size of
+                // the stream to avoid wastefully allocating extra space that will never be needed. It is still
+                // preferable for the user to manually specify the buffer size if it's less than the default, as doing
+                // so allows this branch to be skipped.
+                int fixedBufferSize;
+                try {
+                    fixedBufferSize = inputStream.available();
+                } catch (IOException e) {
+                    // ByteArrayInputStream.available() does not throw.
+                    throw new IllegalStateException(e);
+                }
+                if (alreadyReadLen > 0) {
+                    fixedBufferSize += alreadyReadLen;
+                }
+                if (STANDARD_BUFFER_CONFIGURATION.getInitialBufferSize() > fixedBufferSize) {
+                    configuration = FIXED_SIZE_CONFIGURATIONS[logBase2(fixedBufferSize)];
+                } else {
+                    configuration = STANDARD_BUFFER_CONFIGURATION;
+                }
+            } else {
+                configuration = STANDARD_BUFFER_CONFIGURATION;
+            }
+        }
+        validate(configuration);
+        peekIndex = 0;
+        checkpoint = 0;
+
+        for (int i = 0; i < CONTAINER_STACK_INITIAL_CAPACITY; i++) {
+            containerStack[i] = new Marker(-1, -1);
+        }
```

Comment:
I tried this and benchmarked with our small (~100 byte) dataset where reader construction time dominates. I found the `Arrays.setAll` technique to be up to 1.6% slower than the loop, so I'll keep the loop at this point since I don't think the difference in readability is that significant.